### ISSUE TITLE
Current implementation disallows using GC in destructors

### ISSUE THREAD

**Issue by Manishearth** [Manishearth/rust-gc#7]
```
#![feature(plugin, custom_derive)]

#![plugin(gc_plugin)]
extern crate gc;

use gc::{GcCell, Gc, force_collect};

#[derive(Trace)]
struct Droppable {
    thing: Gc<GcCell<u8>>
}

impl Drop for Droppable {
    fn drop(&mut self) {
        let mut thing = self.thing.borrow_mut();
        *thing = *thing + 1;
    }
}

#[test]
fn test_drop() {
    {
        let x = Gc::new(Droppable {thing: Gc::new(GcCell::new(0))});
    }   
    force_collect()
}
```

Here, the panic on GcBox::value is triggered because drop() is called while GC_SWEEPING is true.

We should probably make GC_SWEEPING a tristate, and set it to the third state when we need to run destructors (the `mem::swap`). In this stage, GC manipulations are allowed, however a sweep cannot be triggered.

Then again, that might not fix the problem, since it's possible to clone the GC in its destructor and pop it into TLS or something. Hmm.

cc @mystor

**Comment by mystor** [comment]
The reason why the GC_SWEEPING is present is because garbage collected objects can be destroyed in cycles. If, for example, Droppable was participating in a cycle, we can't guarantee which object in the cycle is freed first. 

This means that when drop is being run on an object which is being collected, all GC-ed pointers are potentially invalid, as they could be pointing to memory which was Dropped earlier.

**Comment by Manishearth** [comment]
Hmm, right. We could forbid GC manipulation within a drop impl perhaps. (Via a lint and/or declaring it illegal).

**Comment by andersk** [comment]
This seems to have been fixed back in #45.
